### PR TITLE
Remove redundant Mounts when start docker daemon

### PR DESCRIPTION
**Pull request by keloyang**
Signed-off-by: yangshukui <yangshukui@huawei.com>

If you have one or more container running in your host and restart docker daemon when docker save is in progress, then  you use docker save again, the docker save operation will not be successful again.

**- Why**
1. with container ruuning, docker daemon will not invoke Cleanup in the graphdriver when daemon exit.
2. docker daemon exit with exception when docker save operation is running,the procedure of saving will have mount/umount operation , if docker daemon exit with exception, it maybe has something not umount.

**- How I did it**
when restart docker daemon, remove the redundant mounts which are not used by docker container.

**- How to verify it**
before the pr,
```
docker run -d ubuntu:15.04 sleep 100000
docker save ubuntu:15.04>test.tar         #(kill docker daemon at the same time and restart it)
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
docker save rnd-dockerhub.huawei.com/official/ubuntu:15.04>test.tar
Error response from daemon: open /var/lib/docker/devicemapper/mnt/b49e7f17506fc456afc1936ddadc570ff9861f86af5c157b10b07057b7a31535/rootfs/bin/bash: no such file or directory
docker save rnd-dockerhub.huawei.com/official/ubuntu:15.04>test.tar
Error response from daemon: open /var/lib/docker/devicemapper/mnt/b49e7f17506fc456afc1936ddadc570ff9861f86af5c157b10b07057b7a31535/rootfs/bin/bash: no such file or directory
 ......
```

### CODE REVIEW

File: daemon/daemon.go [hunk 1]
```diff
@@ -123,6 +125,35 @@ func (daemon *Daemon) HasExperimental() bool {
...
+		return err
+	}
+	defer f.Close()
+
+	root := fmt.Sprintf("%s/%s", daemon.root, daemon.layerStore.DriverName())
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		text := scanner.Text()
+		fields := strings.Split(text, " ")
+		path := fields[4]
```
Comment (thaJeztah) [inline daemon/daemon.go]: This requires a check if there's 4 fields, otherwise this will panic

**Comment by thaJeztah** [comment]
ping @cpuguy83 ptal

**Comment by coolljt0725** [comment]
ping @cpuguy83  @vdemeester

**Comment by keloyang** [comment]
Can I remove "if len(fields) < 5 {",  “/proc/self/mountinfo” is a os interface which need forward compatibility, so I think there is no need to judge the length.
ping @thaJeztah WDYT ?

**Comment by keloyang** [comment]
ping @cpuguy83 @vdemeester

**Comment by vdemeester** [comment]
@keloyang unit test failures 

```
08:04:52 # github.com/docker/docker/distribution/xfer
08:04:52 distribution/xfer/download_test.go:270: cannot use layerStore (type *mockLayerStore) as type layer.Store in argument to NewLayerDownloadManager:
08:04:52 	*mockLayerStore does not implement layer.Store (missing DriverPut method)
08:04:52 distribution/xfer/download_test.go:331: cannot use mockLayerStore literal (type *mockLayerStore) as type layer.Store in argument to NewLayerDownloadManager:
08:04:52 	*mockLayerStore does not implement layer.Store (missing DriverPut method)
```

File: daemon/daemon.go [hunk 2]
```diff
@@ -123,6 +124,53 @@ func (daemon *Daemon) HasExperimental() bool {
 	return false
 }
 
+func (daemon *Daemon) removeRedundantMounts(containers map[string]*container.Container) error {
```
Comment (AkihiroSuda) [inline daemon/daemon.go]: this should be moved to `daemon_linux.go`

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

**Comment by thaJeztah** [comment]
ping @AkihiroSuda @cpuguy83 PR was updated, PTAL

File: layer/layer_store.go [hunk 3]
```diff
@@ -665,6 +665,10 @@ func (ls *layerStore) Cleanup() error {
 	return ls.driver.Cleanup()
 }
 
+func (ls *layerStore) DriverPut(id string) error {
```
Comment (cpuguy83) [inline layer/layer_store.go]: I don't think we should be adding this to the interface

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
Seems like instead of this the layerstore should have a cleanup function that is run on daemon startup.
Ideally a `kill -9` should never lead to an inconsistent state after startup.

File: layer/layer_store.go (hunk 3 above)
Comment (keloyang) [inline layer/layer_store.go]: >layerstore should have a cleanup function that is run on daemon startup.
Thanks for your review.
how can we get a method to  exclude the correct mounts which are used by the running containers?

Comment (cpuguy83) [inline layer/layer_store.go]: > Seems like instead of this the layerstore should have a cleanup function that is run on daemon startup.

**Pull request closed without merging**